modules/bismark/align/main.nf

```diff
@@ -29,14 +29,46 @@ process BISMARK_ALIGN {
     path "versions.yml"                 , emit: versions
 
     script:
+    // Try to assign sensible bismark memory units according to what the task was given
+    def ccore = 1
+    def cpu_per_multicore = 3
+    def mem_per_multicore = (13.GB).toBytes()
+
+    if( task.cpus ){
+        // Numbers based on recommendation by Felix for a typical mouse genome
+        if( params.single_cell || params.zymo || params.non_directional ){
+            cpu_per_multicore = 5
+            mem_per_multicore = (18.GB).toBytes()
+        }
+        // Check if the user has specified this and overwrite if so
+        if(params.bismark_align_cpu_per_multicore) {
+            cpu_per_multicore = (params.bismark_align_cpu_per_multicore as int)
+        }
+        if(params.bismark_align_mem_per_multicore) {
+            mem_per_multicore = (params.bismark_align_mem_per_multicore as nextflow.util.MemoryUnit).toBytes()
+        }
+        // How many multicore splits can we afford with the cpus we have?
+        ccore = ((task.cpus as int) / cpu_per_multicore) as int
+        // Check that we have enough memory, assuming 13GB memory per instance (typical for mouse alignment)
+        try {
+            def tmem = (task.memory as nextflow.util.MemoryUnit).toBytes()
+            def mcore = (tmem / mem_per_multicore) as int
+            ccore = Math.min(ccore, mcore)
+        } catch (all) {
+            log.warn "Not able to define bismark align multicore based on available memory"
+        }
+    }
+    def multicore  = (ccore > 1) ? "--multicore ${ccore}" : ''
```

Comment:
Why not just `--multicore ${task.cpus}` and leave it to the pipeline developer to assign appropriate memory resources?

Reply:
Because the Bismark flags don't correspond with actual usage. So if you want to use 8 cores then you need to specify a different number on the command line.

The calculation to reach this number is (mostly) specific to Bismark and not pipeline / sample specific. So if we can handle that complexity in the module it standardises in and means the pipeline developer / user only needs to think about cpus used (as is normal).

In theory.